fastestimator/pipeline.py

```diff
@@ -13,442 +13,132 @@
-            op_data = op.forward(op_data, self.state_dict)
-            if op.outputs:
-                write_outputs_by_key(item, op_data, op.outputs)
-        return item
-
-    def __len__(self):
-        return len(self.dataset)
-
-    def set_epoch_and_mode(self, epoch: int, mode: str):
-        self.epoch = epoch
-        if self.op_schedule is not None:
-            self.epoch_ops = self.op_schedule.get_current_value(epoch=epoch)
-            # TODO - rather than filtering every epoch the ops should probably be separated up-front
-            self.epoch_ops = list(filter(lambda op: op.mode is None or op.mode == mode, self.epoch_ops))
-        self.state_dict = {"epoch": epoch, "mode": mode}
-
-    def get_signature_epochs(self) -> List[int]:
-        return self.signature_epochs
-
-
-class OpDataLoader(DataLoader):
-    dataset: OpDataset
-
-    def __init__(self, dataset: OpDataset, batch_size: int, shuffle: bool, num_workers: int, drop_last: bool):
-        # It may be tempting to overwrite the collate function so that it works directly with Tensorflow, but their
-        # default one has some memory management tricks that are difficult to replicate, and since we also support raw
-        # data loaders in the estimator we need to put the type conversion at that level anyways.
-        super().__init__(dataset,
-                         batch_size=batch_size,
-                         shuffle=shuffle,
-                         num_workers=num_workers,
-                         drop_last=drop_last,
-                         worker_init_fn=lambda _: np.random.seed())
-        self.set_epoch_and_mode(epoch=0, mode="train")
-
-    def set_epoch_and_mode(self, epoch: int, mode: str):
-        self.dataset.set_epoch_and_mode(epoch, mode)
-
-
-def get_signature_epochs(val: Union[Scheduler, Iterable[Any]]) -> List[int]:
-    signature_epochs = {0}
-    if isinstance(val, Scheduler):
-        signature_epochs.update(val.epoch_dict.keys())
-    else:
-        for elem in val:
-            if isinstance(elem, Scheduler):
-                signature_epochs.update(elem.epoch_dict.keys())
-    return sorted(signature_epochs)
 
+class Pipeline:
+    def __init__(self,
+                 train_data: Union[Dataset, DataLoader, Scheduler, tf.data.Dataset, None] = None,
+                 eval_data: Union[Dataset, DataLoader, tf.data.Dataset, None] = None,
+                 test_data: Union[Dataset, DataLoader, tf.data.Dataset, None] = None,
+                 batch_size: Union[int, Scheduler, None] = None,
+                 ops: Union[None, NumpyOp, Scheduler, List[Union[NumpyOp, Scheduler]]] = None,
+                 num_process: Union[int, None] = None):
+        self.data = {x: y for (x, y) in zip(["train", "eval", "test"], [train_data, eval_data, test_data]) if y}
+        self.batch_size = batch_size
+        self.ops = ops
+        self.num_process = num_process
+        self.fe_dataset_exist = False
```

Comment:
since this is only used in validation step let's not make it a member variable

Reply:
you mean `self.fe_dataset_exist` ?   it has to be member variable because it is needed on scenarios like: `train_data` is data loader, `eval_data` is dataset.